fenicsadapter/adapter_core.py

```diff
@@ -81,39 +83,42 @@ def filter_point_sources(point_sources, filter_out):
     return filtered_point_sources
 
 
-def convert_fenics_to_precice(data, sample_points):
+def convert_fenics_to_precice(function, coupling_subdomain, fenics_dimensions, dimensions):
     """
     Converts data of type dolfin.Function into Numpy array for all x and y coordinates on the boundary.
 
     Parameters
     ----------
-    data : FEniCS function
+    function : FEniCS function
         A FEniCS function referring to a physical variable in the problem.
-    sample_points : array_like
-        The coordinates of the vertices in a numpy array [N x D] where
-        N = number of vertices and D = dimensions of geometry.
 
     Returns
     -------
     array : array_like
         Array of FEniCS function values at each point on the boundary.
     """
-    if type(data) is dolfin.Function:
-        x_all, y_all = sample_points[:, 0], sample_points[:, 1]
-        return np.array([data(x, y) for x, y in zip(x_all, y_all)])
+    _, _, _, _, local_ids, _ = get_coupling_boundary_vertices(function.function_space().mesh(), function.function_space(),
+                                                              coupling_subdomain, fenics_dimensions, dimensions)
```

Comment:
I tried to come up with an approach that helps us to deal with any kind of `function` a user provides. Note that the suggested implementation below still has some errors. I assume that I did not use the different `local_ids` correctly.

# Suggested Implementation

We will have to first transform the `function` (might be anything vector/scalar, low/high order) to a first order function:
```
if function.function_space().num_sub_spaces() > 0:  # function space is a VectorFunctionSpace
    linear_space = VectorFunctionSpace(function.function_space().mesh(), "P", 1)
else:
    linear_space = FunctionSpace(function.function_space().mesh(), "P", 1)

interpolating_function = interpolate(function, linear_space)
```
This makes `interpolating_function.vector().get_local()` easy to understand.

After that, we can `reshape` to get an array that is consistent with the format expected by preCICE:
```
if function.function_space().num_sub_spaces() > 0:  # function space is a VectorFunctionSpace
    assert((len(interpolating_function.function_space().tabulate_dof_coordinates()) / dimensions).is_integer())
    n_vertices = int(len(interpolating_function.function_space().tabulate_dof_coordinates()) / dimensions)
    func_vector = func_vector.reshape([n_vertices, dimensions])
else:  # function space is a FunctionSpace
    n_vertices = len(projected.function_space().tabulate_dof_coordinates())
    func_vector = func_vector.reshape([n_vertices, 1])
```
Finally: Sample `func_vector` where it is necessary:
```
return np.array([func_vector[lid, :] for lid in local_ids])
```

This function should return an array of the correct shape. Compare:
https://github.com/precice/fenics-adapter/blob/666d2dd9d941e4d76ba847cfb0bc6d5e94e3986b/fenicsadapter/adapter_core.py#L103

Reply:
c9e1719 now provides a working version of the approach from above. Note that it works for the serial case only. I was not able to test the parallel case due to my FEniCS installation (which does not support parallel runs).